cmd/thanos/receive.go

```diff
@@ -321,23 +367,145 @@ func runReceive(
 
 		s := shipper.New(logger, reg, dataDir, bkt, func() labels.Labels { return lset }, metadata.ReceiveSource)
 
-		ctx, cancel := context.WithCancel(context.Background())
-		g.Add(func() error {
-			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+		// Before starting, ensure any old blocks are uploaded.
+		if uploaded, err := s.Sync(context.Background()); err != nil {
+			level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+		}
+		// Before quitting, ensure all blocks are uploaded.
+		defer func() {
+			if uploaded, err := s.Sync(context.Background()); err != nil {
+				level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+			}
+		}()
 
-			return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
-				if uploaded, err := s.Sync(ctx); err != nil {
-					level.Warn(logger).Log("err", err, "uploaded", uploaded)
-				}
+		{
+			// Run the uploader in a loop.
+			ctx, cancel := context.WithCancel(context.Background())
+			g.Add(func() error {
+				return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
```

Comment:
seems like we could just write to the "on demand" channel every 30 seconds no?

Reply:
ack sounds nice :)

Reply:
hmm the issue I see with that is that now we have two concurrent writers to the chan and only one of them has control of closing the chan. This is means we will have racy behavior of writing to a closed chan